src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -921,59 +927,55 @@ private string ConsumeTokenToString(int characterCount)
         /// all indeces to the characterBuffer are invalid except for tokenStartIndex.</remarks>
         private bool ReadInput()
         {
-            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
             Debug.Assert(this.tokenStartIndex >= 0 && this.tokenStartIndex <= this.storedCharacterCount, "The token start is out of stored characters range.");
 
             if (this.endOfInputReached)
             {
                 return false;
             }
 
+            // initialze the buffer
+            if (this.characterBuffer == null)
+            {
+                this.characterBuffer = BufferUtils.RentFromBuffer(ArrayPool, InitialCharacterBufferSize);
+            }
+            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
+
             // We need to make sure we have more room in the buffer to read characters into
             if (this.storedCharacterCount == this.characterBuffer.Length)
             {
                 // No more room in the buffer, move or grow the buffer.
                 if (this.tokenStartIndex == this.storedCharacterCount)
                 {
-                    // If the buffer is empty (all characters were consumed from it)
-                    // just start over.
+                    // If the buffer is empty (all characters were consumed from it), just start over.
                     this.tokenStartIndex = 0;
                     this.storedCharacterCount = 0;
                 }
-                else if (this.tokenStartIndex > (this.characterBuffer.Length - MaxCharacterCountToMove))
-                {
-                    // Some characters were consumed, we can just move them in the buffer
-                    // to get more room without allocating.
-                    Array.Copy(
-                        this.characterBuffer,
-                        this.tokenStartIndex,
-                        this.characterBuffer,
-                        0,
-                        this.storedCharacterCount - this.tokenStartIndex);
-                    this.storedCharacterCount -= this.tokenStartIndex;
-                    this.tokenStartIndex = 0;
-                }
-                else
+                else if (this.tokenStartIndex < this.characterBuffer.Length / 2)

```

Comment:
@KanishManuja-MS can you share more detail?

Reply:
What I am trying to say is that this is a good strategy up to a certain limit. Let's say hypothetically, your buffer grows to be very large ( say 1GB), and you run out of the buffer. Instead of allocating another 1 GB maybe we should be allocating only an additional fixed smaller sized memory. I don't want to hold up this PR as this is an edge case  and Mike's large object streaming would be recommended for such cases but I wanted to bring this in your notice.

Reply:
Perhaps we should start out a little less aggressive.  Doubling when the buffer usage is greater than 50% means that, immediately after we double the buffer, and before we return the old one, we could have allocated up to 6x the longest string (if the longest string was 1 character greater than half the buffer at the time of doubling). If we doubled when the buffer usage is greater than 75%, in the worse case scenario we would have temporarily allocated up to 4x the longest string.

---
In reply to: [254012290](https://github.com/OData/odata.net/pull/1388#discussion_r254012290) [](ancestors = 254012290,253684259)